File: 1278-Palindrome Partitioning III.py
class Solution(object):
    def palindromePartition(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        n = len(s)

        def changeCount(left, right):
            count = 0
            while left < right:
                if s[left] != s[right]:
                    count += 1
                left+=1
                right-=1
            
            return count
        
        memo = {}
        
        def dfs(start, kLeft):
            
            if (start, kLeft) in memo:
                return memo[(start, kLeft)]
            
            if kLeft == 1:
                return changeCount(start, n-1)
            
            if n-start == kLeft:
                return 0
            
            totalsplit = float('inf')
            
            for split in range(start, n-kLeft+1):
                total = changeCount(start, split) + dfs(split+1, kLeft-1)
                totalsplit = min(total, totalsplit)

            memo[(start, kLeft)] = totalsplit
            return totalsplit
            
        return dfs(0, k)

File: JavaScript/1278-Palindrome Partitioning III.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var palindromePartition = function(s, k) {
    const n = s.length;
    
    let changeCount = (left, right) =>{
        let count = 0;
        while (left < right){
            if (s[left] !== s[right]){
                count++;
            }
            left++;
            right--;
        }
        return count;
    }
    
    const memo = new Array(n).fill().map(a => new Array(k + 1))
    
    let dfs = (start, kLeft)=>{
        if (memo[start][kLeft] !== undefined){
            return memo[start][kLeft];
        }
        if (n-start === kLeft){
            return 0;
        }
        if (kLeft === 1){
            return changeCount(start, n-1);
        }
        let minimumChange = Number.POSITIVE_INFINITY;
        for (let split = start; split <= n-kLeft; split++){
            minimumChange = Math.min(minimumChange, changeCount(start, split) + dfs(split+1, kLeft-1));
        }
        memo[start][kLeft] = minimumChange;
        return minimumChange;
    }
    return dfs(0, k)
};